packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py

```diff
@@ -0,0 +1,218 @@
+                repeated=configuration.repeated,
+                default_value=None,
+                annotations=dict(configuration.annotations.items()),
+                message_type=configuration.message_type,
+            )
+        )
+
+    output_parameter_list = []
+    for output in metadata.measurement_signature.outputs:
+        output_parameter_list.append(
+            ParameterMetadata.initialize(
+                display_name=output.name,
+                type=output.type,
+                repeated=output.repeated,
+                default_value=None,
+                annotations=dict(output.annotations.items()),
+                message_type=output.message_type,
+            )
+        )
+
+    serialization_descriptors.create_file_descriptor(
+        input_metadata=configuration_parameter_list,
+        output_metadata=output_parameter_list,
+        service_name=service_class,
+        pool=descriptor_pool.Default(),
+    )
+    configuration_metadata = frame_metadata_dict(configuration_parameter_list)
+    deserialized_parameters = decoder.deserialize_parameters(
+        configuration_metadata,
+        metadata.measurement_signature.configuration_defaults.value,
+        service_class + ".Configurations",
+    )
+
+    default_values = [None] * len(deserialized_parameters)
+    for k, v in deserialized_parameters.items():
+        configuration_metadata[k] = configuration_metadata[k]._replace(default_value=v)
+        default_values[k - 1] = v
+
+    return configuration_metadata
+
+
+def get_output_metadata_by_index(
+    metadata: v2_measurement_service_pb2.GetMetadataResponse,
+) -> Dict[int, ParameterMetadata]:
+    """Returns the output metadata of the measurement."""
+    output_parameter_list = []
+    for output in metadata.measurement_signature.outputs:
+        output_parameter_list.append(
+            ParameterMetadata.initialize(
+                display_name=output.name,
+                type=output.type,
+                repeated=output.repeated,
+                default_value=None,
+                annotations=dict(output.annotations.items()),
+                message_type=output.message_type,
+            )
+        )
+    output_metadata = frame_metadata_dict(output_parameter_list)
+    return output_metadata
+
+
+def _remove_invalid_characters(input_string: str, new_char: str) -> str:
+    # Replace any spaces or special characters with an '_'.
+    if not input_string.isidentifier():
+        for invalid_char in INVALID_CHARS:
+            input_string = input_string.replace(invalid_char, new_char)
+
+    if input_string[0].isdigit() or keyword.iskeyword(input_string):
+        input_string = "_" + input_string
+
+    return input_string
+
+
+def get_python_module_name(input_string: str) -> str:
+    """Returns a valid name python module."""
+    module_name = input_string.replace(" ", "_").lower()
+    module_name = _remove_invalid_characters(module_name, "")
+    return module_name
```

Comment:
This loses the hierarchy and crams everything together. 
![image](https://github.com/user-attachments/assets/afa7d0c1-b8fd-4259-8fda-99ab10c5ce48)

Maybe you did this because PEP8 discourages underscores in module names. If so:
- Maybe you should discard the hierarchy and just use the last part of the name.
- Suffixing with `_client` contracts this rationale.

Reply:
Yes. If we add the last part of the name by discarding the hierarchy, say if, user has service class named `sampleMeasurement.Python", If we consider from the period and discard the previous values. This may be a corner case.
So,  I have updated the naming to replace the invalid characters to `"_"` `eg: ni_tests_testmeasurement_python`.
May I know your suggestions?

Reply:
> Yes. If we add the last part of the name by discarding the hierarchy, say if, user has service class named sampleMeasurement.Python, If we consider from the period and discard the previous values. This may be a corner case.

Users should not do that. The documentation should be updated to provide more guidance about how to set the service class. It would make sense to document this in https://www.ni.com/docs/en-US/bundle/measurementplugins/page/python-measurements.html and this repo's README.md files (separate from this PR).

> So, I have updated the naming to replace the invalid characters to "_" eg: ni_tests_testmeasurement_python.
May I know your suggestions?

I do not like including the namespace. Most users will either not use a namespace or will have the same namespace for all of their measurements.

Based on the NI-developed measurements I've seen, many users will copy the "_Python" suffix, so we should strip that if it is present.

So I'd like to see something like this (but tested).
```
base_service_class = service_class.split(".")[-1]
base_service_class = base_service_class.removesuffix("_Python")
if not base_service_class:
    print an error
if not module_name:
    module_name = camel_to_snake_case(base_service_class) + "_client"
if not class_name:
    class_name = base_service_class.replace("_", "") + "Client"
    if not any(ch.isupper() for ch in class_name):
        print a warning that the last part of the service class is expected to be in PascalCase
```
You can copy camel_to_snake_case from https://github.com/ni/nidaqmx-python/blob/master/src/codegen/utilities/helpers.py as a starting point.

Reply:
> * We should document the recommended service class format. We have always named service classes as if they were gRPC services but we neglected to recommend that users do the same.

Sure, adding a new work item for this change and take up in new pull  request.



> * It's okay to throw an error if something is **invalid** but a lowercase class name isn't really invalid.
> * Using `.title()` screws up the existing PascalCase, which makes the class name look worse. `.title()` is for words separated by spaces, not for programmatic identifiers.

- Added a warning to notify the bad style.
- Removed the `'title()`.

Reply:
ok